File: solutions/2026-02-15/cpp/2026-02-15-add-binary.cpp
// Problem: Add Binary
// LeetCode Link: https://leetcode.com/problems/add-binary/
//
// Approach:
// We can simulate binary addition by iterating from the rightmost digits of both strings,
// keeping track of a carry. For each position, we sum the corresponding digits from 'a' and 'b'
// along with the carry. The resulting digit is the sum modulo 2, and the new carry is the sum
// divided by 2. We build the result string in reverse and then reverse it at the end.
//
// Time Complexity: O(max(N, M)), where N and M are the lengths of strings 'a' and 'b'.
// We iterate through the strings at most once.
//
// Space Complexity: O(max(N, M)), for storing the result string.

#include <string>
#include <algorithm>
#include <vector>

class Solution {
public:
    std::string addBinary(std::string a, std::string b) {
        // Initialize the result string.
        std::string result = "";
        // Initialize carry to 0.
        int carry = 0;
        // Pointers to the end of strings a and b.
        int i = a.length() - 1;
        int j = b.length() - 1;

        // Loop while there are digits in either string or there's a carry.
        while (i >= 0 || j >= 0 || carry) {
            // Get the digit from string a, if available, otherwise assume 0.
            int digit_a = (i >= 0) ? (a[i] - '0') : 0;
            // Get the digit from string b, if available, otherwise assume 0.
            int digit_b = (j >= 0) ? (b[j] - '0') : 0;

            // Calculate the sum of the current digits and the carry.
            int current_sum = digit_a + digit_b + carry;

            // The last digit of the sum (modulo 2) is the current result digit.
            // Convert the integer digit back to a character.
            result += std::to_string(current_sum % 2);

            // The carry for the next iteration is the sum divided by 2.
            carry = current_sum / 2;

            // Move to the next digit in string a (if available).
            i--;
            // Move to the next digit in string b (if available).
            j--;
        }

        // The result string was built in reverse order, so reverse it.
        std::reverse(result.begin(), result.end());

        // Return the final binary sum.
        return result;
    }
};

File: solutions/2026-02-15/javascript/2026-02-15-add-binary.js
// Problem: Add Binary
// Given two binary strings, return their sum as a binary string.
// Link: https://leetcode.com/problems/add-binary/
//
// Approach:
// We can simulate the process of binary addition just like we do with decimal numbers,
// from right to left. We'll use a carry variable to keep track of any overflow from
// adding the bits at the current position.
//
// 1. Initialize an empty string `result` to store the sum.
// 2. Initialize `carry` to 0.
// 3. Use two pointers, `i` and `j`, to iterate from the end of strings `a` and `b` respectively.
// 4. Loop while `i` or `j` are non-negative or `carry` is greater than 0.
// 5. In each iteration, get the current bits from `a` and `b` (or 0 if the pointer has gone past the beginning of the string). Convert these characters to integers.
// 6. Calculate the `sum` of the current bits and the `carry`.
// 7. The last bit of the `sum` (sum % 2) is the current bit of our `result`. Prepend this to the `result` string.
// 8. The new `carry` is `Math.floor(sum / 2)`.
// 9. Decrement `i` and `j`.
// 10. After the loop, `result` will contain the binary sum.
//
// Time Complexity: O(max(n, m)), where n is the length of string a and m is the length of string b.
// We iterate through the strings at most once, from right to left.
//
// Space Complexity: O(max(n, m)) for the result string. In the worst case, the sum can be one digit longer than the longer input string.

/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
var addBinary = function(a, b) {
    let result = ""; // Initialize an empty string to store the binary sum
    let carry = 0;   // Initialize carry to 0

    // Initialize pointers to the end of both strings
    let i = a.length - 1;
    let j = b.length - 1;

    // Loop while there are digits in either string or there's a carry
    while (i >= 0 || j >= 0 || carry > 0) {
        // Get the current digit from string a, default to 0 if pointer is out of bounds
        const digitA = i >= 0 ? parseInt(a[i]) : 0;
        // Get the current digit from string b, default to 0 if pointer is out of bounds
        const digitB = j >= 0 ? parseInt(b[j]) : 0;

        // Calculate the sum of the current digits and the carry
        const sum = digitA + digitB + carry;

        // The current bit of the result is the remainder when sum is divided by 2
        // Prepend this bit to the result string
        result = (sum % 2) + result;

        // The new carry is the quotient when sum is divided by 2
        carry = Math.floor(sum / 2);

        // Move the pointers to the next digit (to the left)
        i--;
        j--;
    }

    // Return the final binary sum
    return result;
};
